python/cudf/cudf/_lib/orc.pyx

```diff
@@ -62,9 +73,206 @@ from pyarrow.lib import NativeFile
 from cudf._lib.utils import _index_level_name, generate_pandas_metadata
 
 
-cpdef read_raw_orc_statistics(filepath_or_buffer):
+cdef get_statistics_type(statistics_type stats):
+    index = stats.index()
+    types = {
+        0: "no_statistics",
+        1: "integer_statistics",
+        2: "double_statistics",
+        3: "string_statistics",
+        4: "bucket_statistics",
+        5: "decimal_statistics",
+        6: "date_statistics",
+        7: "binary_statistics",
+        8: "timestamp_statistics",
+    }
+    return types[index]
+
+
+cdef extern from *:
+    """
+    #include <cudf/io/orc_metadata.hpp>
+    namespace cudf::io {
+        integer_statistics get_integer_statistics(statistics_type stats) {
+            return std::get<integer_statistics>(stats);
+        }
+
+        double_statistics get_double_statistics(statistics_type stats) {
+            return std::get<double_statistics>(stats);
+        }
+
+        string_statistics get_string_statistics(statistics_type stats) {
+            return std::get<string_statistics>(stats);
+        }
+
+        bucket_statistics get_bucket_statistics(statistics_type stats) {
+            return std::get<bucket_statistics>(stats);
+        }
+
+        decimal_statistics get_decimal_statistics(statistics_type stats) {
+            return std::get<decimal_statistics>(stats);
+        }
+
+        date_statistics get_date_statistics(statistics_type stats) {
+            return std::get<date_statistics>(stats);
+        }
+
+        binary_statistics get_binary_statistics(statistics_type stats) {
+            return std::get<binary_statistics>(stats);
+        }
+
+        timestamp_statistics get_timestamp_statistics(statistics_type stats) {
+            return std::get<timestamp_statistics>(stats);
+        }
+    }
+    """
+    cdef integer_statistics get_integer_statistics(statistics_type stats)
+    cdef double_statistics get_double_statistics(statistics_type stats)
+    cdef string_statistics get_string_statistics(statistics_type stats)
+    cdef bucket_statistics get_bucket_statistics(statistics_type stats)
+    cdef decimal_statistics get_decimal_statistics(statistics_type stats)
+    cdef date_statistics get_date_statistics(statistics_type stats)
+    cdef binary_statistics get_binary_statistics(statistics_type stats)
+    cdef timestamp_statistics get_timestamp_statistics(statistics_type stats)
+
+
+cdef _parse_column_type_statistics(column_statistics stats):
+    # Initialize stats to return and parse stats blob
+    column_stats = {}
+
+    if stats.number_of_values.has_value():
```

Comment:
I didn't find a good way to improve this. Happy to have further ideas.

Reply:
You can do `value_or`, but I'm guessing it won't work nicely if you pass `None` (or any Python object). If there is a meaningful difference between setting `number_of_values` to 0 vs None (or leaving it out, as you're currently doing) then I don't think there's a good improvement to be made. OTOH if you could just set this to 0 and `has_null` to False by default below, then `value_or` should work.

Reply:
Yeah, the problem here is that I'm trying to match the previous logic exactly. Some keys get `None` values if they don't exist, and other keys are not set if they don't exist. 🤷‍♂️  Since I can't use `value_or(None)`, there's not a better approach here afaict.